f3/granite.go

```diff
@@ -394,69 +394,54 @@ func (i *instance) isJustified(msg *GMessage) error {
 		}
 		return nil
 	}
-
+	// Require justification for all other messages.
 	if msg.Justification == nil {
 		return fmt.Errorf("message for phase %v round %v has no justification", msg.Vote.Step, msg.Vote.Round)
 	}
+	// Always check that the justification is for the same instance.
 	if msg.Vote.Instance != msg.Justification.Vote.Instance {
 		return fmt.Errorf("message with instanceID %v has evidence from instanceID: %v", msg.Vote.Instance, msg.Justification.Vote.Instance)
 	}
 
-	switch msg.Vote.Step {
-	case CONVERGE_PHASE:
-		//CONVERGE is justified by a strong quorum of COMMIT for bottom from the previous round.
-		// or a strong quorum of PREPARE for the same value from the previous round.
-		prevRound := msg.Vote.Round - 1
-		if msg.Justification.Vote.Round != prevRound {
-			return fmt.Errorf("CONVERGE %v has evidence from wrong round %d", msg.Vote.Round, msg.Justification.Vote.Round)
-		}
+	// Check every remaining field of the justification, according to the phase requirements.
+	// This map goes from the message phase to the expected justification phase(s),
+	// to the required vote values.
+	// Anything else is disallowed.
+	expectations := map[Phase]map[Phase]struct {
+		Round uint64
+		Value ECChain
+	}{
+		// CONVERGE is justified by a strong quorum of COMMIT for bottom,
+		// or a strong quorum of PREPARE for the same value, from the previous round.
+		CONVERGE_PHASE: {
+			COMMIT_PHASE:  {msg.Vote.Round - 1, ECChain{}},
+			PREPARE_PHASE: {msg.Vote.Round - 1, msg.Vote.Value},
```

Comment:
The previous code made sure that the `PREPARE` was not for the zero value, but actually that is ok to have, so this is good.